python/frequenz/microgrid_component_graph/__init__.pyi

```diff
@@ -0,0 +1,264 @@
+        """Check if the specified component is an EV charger meter.
+
+        Args:
+            component_id: ID of the component to check.
+
+        Returns:
+            Whether the specified component is an EV charger meter.
+
+        Raises:
+            ValueError: if no component exists with the given ID.
+        """
+
+    def is_chp_meter(self, component_id: ComponentIdProtocol) -> bool:
+        """Check if the specified component is a CHP meter.
+
+        Args:
+            component_id: ID of the component to check.
+
+        Returns:
+            Whether the specified component is a CHP meter.
+
+        Raises:
+            ValueError: if no component exists with the given ID.
+        """
+
+    def consumer_formula(self) -> str:
+        """Generate the consumer formula for this component graph.
+
+        Returns:
+            The consumer formula as a string.
+        """
+
+    def producer_formula(self) -> str:
+        """Generate the producer formula for this component graph.
+
+        Returns:
+            The producer formula as a string.
+        """
+
+    def grid_formula(self) -> str:
+        """Generate the grid formula for this component graph.
+
+        Returns:
+            The grid formula as a string.
+        """
+
+    def pv_formula(self, pv_inverter_ids: abc.Set[ComponentIdProtocol] | None) -> str:
+        """Generate the PV formula for this component graph.
+
+        Returns:
+            The PV formula as a string.
+        """
+
+    def battery_formula(self, battery_ids: abc.Set[ComponentIdProtocol] | None) -> str:
+        """Generate the battery formula for this component graph.
+
+        Returns:
+            The battery formula as a string.
+        """
+
+    def chp_formula(self, chp_ids: abc.Set[ComponentIdProtocol] | None) -> str:
+        """Generate the CHP formula for this component graph.
+
+        Returns:
+            The CHP formula as a string.
+        """
+
+    def ev_charger_formula(
+        self, ev_charger_ids: abc.Set[ComponentIdProtocol] | None
+    ) -> str:
+        """Generate the EV charger formula for this component graph.
+
+        Returns:
+            The EV charger formula as a string.
+        """
+
+    def grid_coalesce_formula(self) -> str:
+        """Generate the grid coalesce formula for this component graph.
+
+        Returns:
+            The grid coalesced formula as a string.
+        """
+
+    def battery_coalesce_formula(
+        self, battery_ids: abc.Set[ComponentIdProtocol] | None
+    ) -> str:
+        """Generate the battery coalesce formula for this component graph.
+
+        Returns:
+            The battery coalesced formula as a string.
+        """
+
+    def pv_coalesce_formula(
+        self, pv_inverter_ids: abc.Set[ComponentIdProtocol] | None
+    ) -> str:
+        """Generate the PV coalesce formula for this component graph.
+
+        Returns:
+            The PV coalesced formula as a string.
+        """
```

Comment:
I guess there is no way to convert a "regular" formula to a coalesce one to avoid the duplicates, right? Also any reasons to only provide coalesce formulas for grid, pv and batteries?

Reply:
> I guess there is no way to convert a "regular" formula to a coalesce one to avoid the duplicates, right?

No, but maybe we can look into that later.

> Also any reasons to only provide coalesce formulas for grid, pv and batteries?

Those were the only ones required for one of our internal apps for which these were added.  I expect only the grid coalesce formula being useful for voltage, frequency streams in the SDK, but might not be too easy, need to look into that as well, later.